data/dataGeneratorScripts/generateRenewables/importMarktStammDaten.py

```diff
@@ -0,0 +1,372 @@
+from collections import defaultdict
+import csv
+from lxml import etree
+import os
+import re
+import json
+
+
+cutoffdate="31.12.2018"
+
+# this is only used in the handling of "not 2018"-ags keys 
+# this is not used in the xml parsing so far
+
+exceptionDict={
+"09478444":"Neuensorger Forst ist ein gemeindefreies gebiet",
+"01053105":"Sachsenwald ist ein gemeindefreies Gebiet",
+"06435200":"Gutsbezirk Spessart ist ist ein gemeindefreies Gebiet",
+"01060014":"BuchholzerForstist ist ein gemeindefreies Gebiet",
+"09572444":"Gdefr. Geb. (Lkr Erlangen-Höchstadt) ist ein gemeindefreies Gebiet",
+"03155501":"Solling (Landkreis Northeim) ist ein gemeindefreies Gebiet",
+"03153504":"Harz (Landkreis Goslar), gemfr. Gebiet ist ein gemeindefreies Gebiet"}
```

Comment:
Could you explain the process by which you reduced the [AGS history](https://www.xrepository.de/api/xrepository/urn:xoev-de:bund:destatis:bevoelkerungsstatistik:codeliste:ags.historie_2021-12-31/download/Destatis.AGS.Historie_2021-12-31.json) to the short list here in `exceptionDict`?

Reply:
@curiousleo wir berechnen mit local zero ja nur die AGS-Schlüssel in unserer Master Liste mit dem stand 31.12.2018. (siehe [localzero-data-public](https://github.com/GermanZero-de/localzero-data-public)/ags/master.csv)
Dort haben wir keine gemeindefreien Gebiete mit aufgenommen, weil wir für die AGS-Schlüssel auch in anderen Daten-Tabellen noch Datenlücken haben. (Gemeindefreie Gebiete sind meistens Naturflächen/Wälder/Seen etc., die keine eigene Verwaltung haben und wo sonst auch nicht sonderlich viel passiert).
Ich prüfe in dem Skript, nachdem ich alle Markstammdatenregister Einheiten, den bei ihnen eingetragenen Ags-Schlüsseln zugeordnet habe, ob alle hieraus resultierenden AGS-Schlüssel auch in der Master Liste sind. Ist das nicht der Fall gibt es verschiedene Optionen: 

- Der Ags-Schlüssel ist ein alter Schlüssel einer kommune, die **vor** Ende 2018 ihren Ags geändert hat.
- Der Ags-Schlüssel ist ein neuer Schlüssel einer kommune, die **nach** Ende 2018 ihren Ags geändert hat.
- Der Ags-Schlüssel gehört zu einem gemeindefreien Gebiet und ist deswegen nicht in der Masterliste.

Aktuell frage ich für alle diese Fälle die json, die man unter https://www.xrepository.de/api/xrepository/urn:xoev-de:bund:destatis:bevoelkerungsstatistik:codeliste:ags.historie_2021-12-31/download/Destatis.AGS.Historie_2021-12-31.json findet, ob es für die jeweiligen Schlüssel Änderungen vor oder nach 2018 gab und passe diese an (wie genau ich das mache ist hier vielleicht erstmal nicht relevant, aber können wir auch gerne nochmal drüber reden).
Danach überprüfe ich nochmal, ob die nun angepassten Schlüssel in der Masterliste sind. 
Dies war selbstverständlich nicht der Fall bei den 7 gemeindefreien Gebieten in meiner Exception list. Die habe ich manuell erstellt mit mehreren Script-Durchläufen. Das Script stoppt, wenn der Fall auftritt, dass ein angepasster AGS nicht in der Master liste enthalten ist.

Hilft dir das? :)

Reply:
@Ultramarin95 danke für die ausführliche Erklärung.

Verstehe ich es richtig, dass die Werte in dem markierten dictionary quasi den Grund angeben, warum dieser AGS ignoriert wird? (Konkret zum Beispiel: "AGS 01053105 wird ignoriert, weil Sachsenwald ein gemeindefreies Gebiet ist".)

Dann halte ich es für eine gute Idee, das dictionary kurz zu kommentieren, etwa mit "Diese AGS werden ignoriert, zum Beispiel weil wir zurzeit für gemeindefreie Gebiete keine Daten aufbereiten. Die den AGS zugeordneten Werte geben den Grund an, warum der jeweilige AGS ignoriert wird."